### PR TITLE
Fix sliding sync on workers

### PR DESCRIPTION
**Pull request by erikjohnston**
Broke in #17630

### CODE REVIEW

File: synapse/storage/databases/main/roommember.py [hunk 1]
```diff
@@ -1424,6 +1425,17 @@ def get_sliding_sync_rooms_for_user_txn(
             get_sliding_sync_rooms_for_user_txn,
         )
 
+    async def have_finished_sliding_sync_background_jobs(self) -> bool:
+        """Return if its safe to use the sliding sync membership tables."""
```
Comment (anoadragon453) [inline synapse/storage/databases/main/roommember.py:1429]: the nittiest of nits.

```suggestion
        """Return if it's safe to use the sliding sync membership tables."""
```

**Review by anoadragon453 — APPROVED** [review]
So the problem was that this background update could only run on the main process, not the background worker?

File: synapse/storage/databases/main/events_bg_updates.py [hunk 2]
```diff
@@ -2342,17 +2342,6 @@ def _fill_table_txn(txn: LoggingTransaction) -> None:
...
-        """Return if its safe to use the sliding sync membership tables."""
-
-        return await self.db_pool.updates.have_completed_background_updates(
-            (
-                _BackgroundUpdates.SLIDING_SYNC_PREFILL_JOINED_ROOMS_TO_RECALCULATE_TABLE_BG_UPDATE,
-                _BackgroundUpdates.SLIDING_SYNC_JOINED_ROOMS_BG_UPDATE,
-                _BackgroundUpdates.SLIDING_SYNC_MEMBERSHIP_SNAPSHOTS_BG_UPDATE,
-            )
-        )
-
```
Comment (MadLittleMods) [inline synapse/storage/databases/main/events_bg_updates.py:2355]: What was the problem on workers in the first place?

How does this fix the problem?

Comment (erikjohnston) [reply synapse/storage/databases/main/events_bg_updates.py:2355]: Not all of the storage classes are used on workers, i.e. there are actually two mixins of data stores: one for workers and one for the master. This store isn't in the worker data store.

Ideally we'd merge the two datastores and add asserts to the functions that can't be run on workers.

**Pull request merged**